File: projects/number_guess/src/game.py
import random

def generate_target(min_val: int = 1, max_val: int = 100) -> int:
    """Generates a random target number within the specified range."""
    return random.randint(min_val, max_val)

def evaluate_guess(secret: int, guess: int) -> str:
    """Compares the guess to the secret number and returns the result message."""
    if guess < secret:
        return "too low"
    elif guess > secret:
        return "too high"
    else:
        return "correct"

def play_round():
    """Handles a single round of the guessing game."""
    target = generate_target()
    print("\nI'm thinking of a number between 1 and 100.")
    
    while True:
        try:
            user_input = input("Enter your guess: ")
            guess = int(user_input)
        except ValueError:
            print("Please enter a valid integer.")
            continue

        result = evaluate_guess(target, guess)
        
        if result == "correct":
            print("Congratulations! You guessed it!")
            break
        else:
            print(f"Your guess was {result}.")

def main():
    """Main entry point for the game loop."""
    print("Welcome to the Number Guessing Game!")
    
    while True:
        play_round()
        
        choice = input("Would you like to play again? (y/n): ").strip().lower()
        if choice != 'y':
            print("Thanks for playing! Goodbye.")
            break

if __name__ == "__main__":
    main()

File: projects/number_guess_go/src/game.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type GuessResult int

const (
	TooLow GuessResult = iota
	TooHigh
	Correct
)

// GenerateTarget returns a random number between 1 and 100.
func GenerateTarget() int {
	// Using a local random source to avoid global state issues if needed,
	// but for this simple implementation, we'll use the global one seeded in main.
	return rand.Intn(100) + 1
}

// CompareGuess compares the guess with the target and returns the result.
func CompareGuess(target, guess int) GuessResult {
	if guess < target {
		return TooLow
	} else if guess > target {
		return TooHigh
	}
	return Correct
}

// PlayRound handles the I/O for a single round of the game.
func PlayRound(target int) error {
	fmt.Printf("\nI'm thinking of a number between 1 and 100.\n")

	for {
		var guess int
		fmt.Print("Enter your guess: ")
		_, err := fmt.Scan(&guess)
		if err != nil {
			return fmt.Errorf("invalid input: please enter an integer")
		}

		result := CompareGuess(target, guess)
		switch result {
		case TooLow:
			fmt.Println("Too low! Try again.")
		case TooHigh:
			fmt.Println("Too high! Try again.")
		case Correct:
			fmt.Printf("Congratulations! You guessed it: %d\n", target)
			return nil
		}
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())

	for {
		target := GenerateTarget()
		err := PlayRound(target)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
			break
		}

		var choice string
		fmt.Print("Would you like to play again? (y/n): ")
		fmt.Scan(&choice)
		if choice != "y" && choice != "Y" {
			fmt.Println("Thanks for playing!")
			break
		}
	}
}
